XR_APILAYER_NOVENDOR_toolkit/fsr.cpp

```diff
@@ -0,0 +1,216 @@
+            if (!m_intermediary) {
+                const auto& infos = output->getInfo();
+                initializeIntermediary(infos.width, infos.height, 0 /* infos.format */);
+            }
+
+            if (!m_isSharpenOnly) {
+                m_device->setShader(m_shaderEASU);
+                m_device->setShaderInput(0, m_configBuffer);
+                m_device->setShaderInput(0, input, slice);
+                m_device->setShaderOutput(0, m_intermediary);
+                m_device->dispatchShader();
+            }
+
+            m_device->setShader(m_shaderRCAS);
+            m_device->setShaderInput(0, m_configBuffer);
+            m_device->setShaderInput(0, m_isSharpenOnly ? input : m_intermediary);
+            m_device->setShaderOutput(0, output, slice);
+            m_device->dispatchShader();
+        }
+
+      private:
+        void initializeScaler() {
+            const auto shadersDir = std::filesystem::path(dllHome) / std::filesystem::path("shaders");
+            const auto shaderPath = shadersDir / std::filesystem::path("FSR.hlsl");
+
+            // This value is the image region dimension that each thread group of the FSR shader operates on
+            const auto threadGroupWorkRegionDim = 16u;
+            const std::array<unsigned int, 3> threadGroups = {
+                (m_outputWidth  + (threadGroupWorkRegionDim - 1)) / threadGroupWorkRegionDim, // dispatchX
+                (m_outputHeight + (threadGroupWorkRegionDim - 1)) / threadGroupWorkRegionDim, // dispatchY
+                1 
+            };
+
+            // EASU/RCAS common
+            utilities::shader::Defines defines;
+            defines.add("FSR_THREAD_GROUP_SIZE", 64);
+            defines.add("SAMPLE_SLOW_FALLBACK", 1);
+            defines.add("SAMPLE_BILINEAR", 0);
+            defines.add("SAMPLE_HDR_OUTPUT", 0);
+
+            // EASU specific
+            defines.add("SAMPLE_RCAS", 0);
+            defines.add("SAMPLE_EASU", 1);
+            m_shaderEASU = m_device->createComputeShader(
+                shaderPath.string(), "mainCS", "FSR EASU CS", threadGroups, defines.get(), shadersDir.string());
+
+            // RCAS specific
+            defines.set("SAMPLE_EASU", 0);
+            defines.set("SAMPLE_RCAS", 1);
+            m_shaderRCAS = m_device->createComputeShader(
+                shaderPath.string(), "mainCS", "FSR RCAS CS", threadGroups, defines.get(), shadersDir.string());
+
+            m_isSharpenOnly = false;
+        }
+
+        void initializeSharpen() {
+            initializeScaler();
+
+            // TODO
+
+            //m_isSharpenOnly = true;
+        }
+
+        void initializeIntermediary(uint32_t width, uint32_t height, int64_t format) {
+            if (format == 0) {
+                // good balance between visuals and perf
+                format = m_device->getTextureFormat(TextureFormat::R16G16B16A16_UNORM);
```

Comment:
Can you confirm this is intentional? I think you had previously mentioned using R10G10B10A2 for the best results.

Reply:
Also, maybe consider removing the `format` argument entirely, since this method is private to this implementation. We can then remove the sRGB check below.

Reply:
This is WIP for now with the idea to keep this generic enough in the future and to create the intermediary texture with the same format as the one already set for the output texture (see line 102). For FSR I find R16G16B16 is the minimum to avoid posterization for now, this is why I'm overriding the format in this commit. But it might be different for NIS or for other post-processing layers we'll add in the future, or when we reach final conclusions about using sRGB format directly. I suggest we keep this approach open and generic for the time being.

Reply:
OK that sounds good.